subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -177,6 +186,19 @@ public void setToolChain(JavaToolChain toolChain) {
         return javaCompiler;
     }
 
+    /**
+     * Obtain a {@link JavaCompiler} matching the {@link JavaToolchainSpec} which can then be used to configure the compiler used by this task.
+     *
+     * @param action The action to configure the {@code JavaToolchainSpec}
+     * @return A {@code Provider<JavaCompiler>}
+     *
+     * @since 6.7
+     */
+    @Incubating
+    public Provider<JavaCompiler> toolchainCompiler(Action<? super JavaToolchainSpec> action) {
```

Comment:
One aspect I find hard to grasp from an implementation POV is that we now have the consumer of the tool (this task) be the same as the producer of the tool (query service/factory exposes on this task). This also will lead to a lot of duplication across plugins as every plugin should/would copy this exact code to expose the same facilities.

How about we split those responsibilities instead? The task will remain the consumer of the tool (keep the compiler property) but move the service into a service that is available as project extension:

```
project.getExtensions().add(JavaCompilerQueryService.class, "toolchainCompiler", someInstance);
```

That way every task can easily configured with just a property the a factory is available project-wide.

Reply:
I don't think we should so this. Extensions are not really meant to expose services. In fact we discussed this approach. I disagree that "the producer is the same as the consumer". It's actually a task which needs to expose the service so that users can configure a value. I makes sense to me that the task exposes this as a method.
Extensions have this huge drawback that they would make this method visible even if you don't care, on a DSL part that has nothing to do with the task.

Reply:
Interesting. I was just recommending what Adam did recently in a very [similar situation](https://github.com/gradle/gradle/commit/2c05da10f1173c172f67e18e2830a4bc87319ff7). Anyway, Louis and I had another idea that he's posting soon to make this a little cleaner

Reply:
I just had a conversation with Louis, I think there's a difference between what Adam did and what you suggest: in your case you're just exposing a service, not more. It's not really _data_ like installations.
I think the approach of using custom property types is actually the right way to go: it's simpler for the user while avoiding exposing unnecessary stuff. Implementation-wise, I have no idea what the cost is.

Reply:
Can you elaborate on this a bit? I don't quite understand how different those two services really are (one is giving us a toolchain/installation based on some criteria (spec), the other one is giving us an installation based on a specific criteria (java home path))

Reply:
After further discussions, we will expose a `javaToolchains` which will be the entry point for getting a tool based on a `JavaToolchainSpec`.